### PR TITLE
SOQL: Support date_n_literal

### PR DESCRIPTION
**Pull request by greg-finley**
### Brief summary of the change made

Closes https://github.com/sqlfluff/sqlfluff/issues/3321

### Are there any other side effects of this change that we should be aware of?
It's possible it goes somewhere besides `Expression_C_Grammar`.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_soql.py [hunk 1]
```diff
@@ -3,18 +3,16 @@
 https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql.htm
 """
 
+from sqlfluff.core.parser.grammar.anyof import OneOf
+from sqlfluff.core.parser.grammar.sequence import Sequence
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_soql.py:7]: We normally just import these from `sqlfluff.core.parser` in other dialects do we not need to specify each one separately.

File: src/sqlfluff/dialects/dialect_soql.py [hunk 2]
```diff
@@ -41,11 +39,53 @@
     "NEXT_FISCAL_YEAR",
 ]
 
-soql_dialect.sets("reserved_keywords").update(date_literals)
+date_n_literals = [
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_soql.py:42]: We use sets for other lists of date times. Any reason for straight arrays here? should we dtandardise on sets to match other dialects?

File: src/sqlfluff/dialects/dialect_soql.py [hunk 3]
```diff
@@ -41,11 +39,53 @@
...
+    "NEXT_N_QUARTERS",
+    "LAST_N_YEARS",
+    "NEXT_N_YEARS",
+    "LAST_N_FISCAL_QUARTERS",
+    "NEXT_N_FISCAL_QUARTERS",
+    "LAST_N_FISCAL_YEARS",
+    "NEXT_N_FISCAL_YEARS",
+]
+
+soql_dialect.sets("reserved_keywords").update(date_literals + date_n_literals)
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_soql.py:59]: Are these definitely reserved keywords? Or they can be used as column names in other dialects so not reserved,  but not sure if that's true here.

Comment (greg-finley) [reply src/sqlfluff/dialects/dialect_soql.py]: I would treat as reserved for now (though I'm fine to swap it if you prefer). I am not seeing documentation either way and don't have access to a Salesforce instance to test it.

**Pull request merged**